linkerd/app/core/src/errors/respond.rs

```diff
@@ -174,150 +286,60 @@ where
-    }
-}
-
-// === helpers ===
-
-fn set_l5d_proxy_error_header(
-    builder: http::response::Builder,
-    error: &(dyn std::error::Error + 'static),
-) -> http::response::Builder {
-    if error.is::<ResponseTimeout>() {
-        builder.header(
-            L5D_PROXY_ERROR,
-            HeaderValue::from_static("request timed out"),
-        )
-    } else if error.is::<ConnectTimeout>() {
-        builder.header(
-            L5D_PROXY_ERROR,
-            HeaderValue::from_static("failed to connect"),
-        )
-    } else if let Some(e) = error.downcast_ref::<FailFastError>() {
-        builder.header(
-            L5D_PROXY_ERROR,
-            HeaderValue::from_str(&e.to_string()).unwrap_or_else(|error| {
-                warn!(%error, "Failed to encode fail-fast error message");
-                HeaderValue::from_static("service in fail-fast")
-            }),
-        )
-    } else if let Some(HttpError { source, .. }) = error.downcast_ref() {
-        builder.header(
-            L5D_PROXY_ERROR,
-            HeaderValue::from_str(&source.to_string())
-                .unwrap_or_else(|_| HeaderValue::from_static("an error occurred")),
-        )
-    } else if let Some(source) = error.source() {
-        set_l5d_proxy_error_header(builder, source)
-    } else {
-        builder.header(
-            L5D_PROXY_ERROR,
-            HeaderValue::from_static("proxy received invalid response"),
-        )
-    }
-}
-
-fn set_http_status(
-    builder: http::response::Builder,
-    error: &(dyn std::error::Error + 'static),
-) -> http::response::Builder {
-    if error.is::<ResponseTimeout>() || error.is::<ConnectTimeout>() {
-        builder.status(StatusCode::GATEWAY_TIMEOUT)
-    } else if error.is::<FailFastError>() {
-        builder.status(StatusCode::SERVICE_UNAVAILABLE)
-    } else if let Some(HttpError { http_status, .. }) = error.downcast_ref() {
-        builder.status(http_status)
-    } else if let Some(source) = error.source() {
-        set_http_status(builder, source)
-    } else {
-        builder.status(StatusCode::BAD_GATEWAY)
-    }
-}
-
-fn set_grpc_status(
-    error: &(dyn std::error::Error + 'static),
-    headers: &mut http::HeaderMap,
-) -> grpc::Code {
-    const GRPC_STATUS: &str = "grpc-status";
-    const GRPC_MESSAGE: &str = "grpc-message";
-
-    if error.is::<ResponseTimeout>() {
-        let code = Code::DeadlineExceeded;
-        headers.insert(GRPC_STATUS, code_header(code));
-        headers.insert(GRPC_MESSAGE, HeaderValue::from_static("request timed out"));
-        code
-    } else if let Some(e) = error.downcast_ref::<FailFastError>() {
-        let code = Code::Unavailable;
-        headers.insert(GRPC_STATUS, code_header(code));
-        headers.insert(
+impl<R, B> ResponseBody<R, B> {
+    fn grpc_trailers(code: tonic::Code, message: &str) -> http::HeaderMap {
+        debug!(grpc.status = ?code, "Synthesizing gRPC trailers");
+        let mut t = http::HeaderMap::new();
```

Comment:
nit/tioli: could be
```suggestion
        let mut t = http::HeaderMap::new(2);
```
to be explicit, although i am pretty sure the first insert will reserve enough capacity for at least 2 headers anyway...